### ISSUE TITLE
Having trouble with reproducing

### ISSUE THREAD

**Issue by jc5201** [JunyiZhu-AI/R-GAP#1]
Hello, I'm trying to reproduce your results.
I'm using your code without any change.

When I run command below, I got image below.
``` python main.py -i 6 ```
![image](https://user-images.githubusercontent.com/9334674/109810054-65bb0680-7c6c-11eb-85d4-3cdf3e61aaab.png)

The result image is quite different from figures of paper.
I expected 2~3 noise dots, but there are lots of noise dots.

Please let me know if there is something I'm missing or something I have to do.

**Comment by JunyiZhu-AI** [comment]
Hello, thanks for your interest in our approach. I just cloned a fresh repo and repeated this experiment three times. They all have the same result as shown below.
![reconstructed](https://user-images.githubusercontent.com/39190294/109812741-b9a5fb00-7c2c-11eb-9ed2-7591463ff7b9.png)

Since the random seed of NumPy and PyTorch is set to zero, we should get the same result. Anyway, such imperfect reconstruction is usually caused by the numerical error from forward and backward propagation as well as the least square solver. For example, if the network is initialized with different parameters, the constraint matrix A and its condition number are different for the same picture, which means the quality of reconstruction may vary, even though this is a closed-form approach.

**Comment by jc5201** [comment]
Which version of pytorch/numpy did you use?
I'll try with that version.

**Comment by JunyiZhu-AI** [comment]
I'm using PyTorch 1.7.1 and NumPy 1.19,5.

**Comment by JunyiZhu-AI** [comment]
I have verified your result on other devices. So far I think this issue is device-dependent, please check these links:
https://stackoverflow.com/questions/53626778/causes-of-floating-point-non-determinism-including-numpy
https://github.com/numpy/numpy/issues/12465
Besides, the least square solver is also regarded as non-deterministic, you might be interested in this:
https://github.com/scipy/scipy/issues/8208

Seems like setting random seed does not necessarily mean a reproducible result, I would like to suggest you comment out the random seed code, you should probably see a much better reconstruction.

**Comment by jc5201** [comment]
Yes. After commenting out random seed, I got much clear results.
It is interesting that It seems like closed-form solution but random seed is important.
Thanks for your help.

Besides, I have one more question.
In my guess, non-deterministic parts are pseudo-inverse, circulant matrix, non-monotonic loss function and floating point. Is it right?

**Comment by JunyiZhu-AI** [comment]
My pleasure. Changing the random seed is just a fast way for you to reproduce the result in our paper, but it doesn't mean that a random seed is important to a closed-form approach. Whereas, the condition number of constraint matrix A is important, as it indicates the numerical stability in our case.

Theoretically, the reconstruction result should be exactly the same with the same random seed, e.g. repeat the experiment on the same machine usually gives out the same result. I suppose that the different numerical errors produced by different machines caused non-consistency between our results. There is a speed and accuracy trade-off in computation. Computers boost calculation speed by losing some accuracy and this is dependent on some low-level libraries. If a numerical error produced by one computer locates close to the direction with a higher singular value then the quality of reconstruction should be worse. So this and the non-deterministic of lst solver are computational issues, but pseudo-inverse, circulant matrix, and loss functions are mathematical concepts, they are deterministic.

**Comment by JunyiZhu-AI** [comment]
Closing this issue, please reopen if there is any further question on this issue.
